File: ml-service/app/utils/logger.py
import json
import sys
import traceback
from datetime import datetime, timezone
from typing import Any


def log(level: str, message: str, **fields: Any) -> None:
    """One JSON line per event. Errors go to stderr. Never pass secrets or request bodies in `fields`."""
    line = json.dumps(
        {"level": level, "time": datetime.now(timezone.utc).isoformat(), "message": message, **fields},
        default=str,
    )
    print(line, file=sys.stderr if level == "error" else sys.stdout, flush=True)


def describe_error(exc: BaseException, include_traceback: bool = True) -> dict[str, Any]:
    """Turns an exception into a loggable dict, optionally with the full traceback."""
    info: dict[str, Any] = {"name": type(exc).__name__, "message": str(exc)}
    if include_traceback:
        info["traceback"] = "".join(traceback.format_exception(exc))
    return info

File: backend/src/utils/logger.ts
type Level = 'info' | 'warn' | 'error';

// Minimal structured logger: one JSON line per event. No logging library is used in Phase 2.
// Logs stay on the server, so they may contain full error details, but never put secrets,
// request bodies or personal data in `fields` on purpose.
export function log(level: Level, message: string, fields: Record<string, unknown> = {}): void {
  const line = JSON.stringify({ level, time: new Date().toISOString(), message, ...fields });
  if (level === 'error') {
    console.error(line);
  } else {
    console.log(line);
  }
}

// Turns anything that was thrown into a loggable object, including the stack trace.
export function describeError(err: unknown): Record<string, unknown> {
  if (err instanceof Error) {
    return { name: err.name, message: err.message, stack: err.stack };
  }
  return { value: String(err) };
}
